File: November/15/kthsmallestinBST.py
# LeetCode 230: Kth Smallest Element in a BST
# Explanation:
# 1. A BST's inorder traversal produces values in sorted order.
# 2. Perform an inorder traversal until we reach the k-th element.
#
# Time Complexity: O(h + k) where h = tree height
# Space Complexity: O(h) for recursion stack

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack = []
        curr = root

        while True:
            while curr:
                stack.append(curr)
                curr = curr.left

            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val

            curr = curr.right

    def kthSmallestRecursive(self, root: TreeNode, k: int) -> int:
        def inorder(node):
            if node:
                yield from inorder(node.left)
                yield node.val
                yield from inorder(node.right)

        gen = inorder(root)
        for _ in range(k - 1):
            next(gen)
        return next(gen)

File: November/15/kthsmallestinBST.go
// LeetCode 230: Kth Smallest Element in a BST
// Explanation:
// Inorder traversal yields nodes in sorted order.
// Count nodes until the k-th is reached.
//
// Time Complexity: O(h + k)
// Space Complexity: O(h)

package main

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
    stack := []*TreeNode{}
    curr := root

    for {
        for curr != nil {
            stack = append(stack, curr)
            curr = curr.Left
        }
        curr = stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        k--
        if k == 0 {
            return curr.Val
        }

        curr = curr.Right
    }
}

